### DISCUSSION TITLE
Multiple DagRuns for the same day after changing the DAG schedule and re-running old tasks.

### DISCUSSION THREAD

**Discussion by Jorricks** [apache/airflow#24187] [General]
### Apache Airflow version

2.2.3

### What happened

In our use-case we often back-fill DAGs(using the clear functionality of DAGRuns) and sometimes have to change DAG schedules. \
This used to be very smooth and without problems, however, we noticed that something changed in the recent versions.

Imagine you have a long running DAG with a schedule of `0 4 * * *`. \
When you clear an old DagRuns of this DAG, it will simply re-run all the tasks and after that continue with the most recent ones. No new DagRuns will be spawned.

Now you change the DAG schedule to `0 2 * * *`. \
When you now clear an old DagRun of this DAG (which was scheduled at 04:00AM), it will re-run the tasks of that DAGRun. However, once its finished, it's starts to spawn DagRuns from that DagRun onwards for 02:00AM.
This means that you end up with two DAGRuns a day now.

I did a bit of snooping in the source code and found out that this is most likely due to this line:
https://github.com/apache/airflow/blob/4fa718e4db2daeb89085ea20e8b3ce0c895e415c/airflow/jobs/scheduler_job.py#L1153

This topic is mostly to start an argument whether this is the intended way or whether it should be as mentioned below.

### What you think should happen instead

No matter if the DAG schedule changed, Airflow should not spawn new dag runs older than the newest (automatically) scheduled DagRun.

### How to reproduce

1. Create a DAG with schedule `0 4 * * *`, `catchup=True` and `start_datetime=datetime.datetime(2022,1,1)`.
2. Enable the DAG and let it catch up.
3. Once caught up, change the DAG schedule and verify it is parsed in the web ui.
4. Clear an old DAGRun.

### Operating System

Tested on Linux & Mac

### Deployment

Other

### Are you willing to submit PR?

- [X] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/apache/airflow/blob/main/CODE_OF_CONDUCT.md)

**Comment by boring-cyborg** [comment]
Thanks for opening your first issue here! Be sure to follow the issue template!

**Comment by uranusjr** [comment]
I don’t know, I can see an argument the current behaviour makes sense for certain cases though. Would it work if when you change the schedule, you also change `start_date`?

**Comment by uranusjr** [comment]
Removing _bug_ for now since it is less of a bug but a… quirk? The behaviour needs to be discussed in more detail.

**Comment by Jorricks** [comment]
> Removing _bug_ for now since it is less of a bug but a… quirk? The behavior needs to be discussed in more detail.

Yes agreed that makes sense.


> I don’t know, I can see an argument the current behaviour makes sense for certain cases though. Would it work if when you change the schedule, you also change `start_date`?

I haven't tested this.
It's a bit odd but one of our users mentioned this as a work around:
1. Manually pause DAG
2. Set DAG runs from success to running via the List Dag Run overview for all relevant backfill dates.
3. Clear all the DagRuns you want to clear
4. Manually un-pause DAG

However, I still feel like this is an unwanted quirk.

**Comment by potiuk** [comment]
Actually this is something that should be solved with Dag Versioning https://cwiki.apache.org/confluence/display/AIRFLOW/AIP-36+DAG+Versioning.  Versioning Schedule is part of this effort too. I am moving that to a discussion - but this topic should be raised when we will start discussing AIP-36
